README.md
# Wow Addons for WLK (3.3.5)
基于多玩魔盒 3.3.5, 添加自用单体插件整合而成

## 插件列表
名称                             | 描述
------------------------------- | ---------------------------------
!BugGrabber                     | BUG 收集工具
BugSack                         | BUG 显示插件
_NPCScan                        | 稀有精英探测+地图路径
Accountant_Classic              | [多玩] 收支统计
AckisRecipeList                 | [商业] 全商业技能信息列表
ActionBar_Customer              | [界面] 动作条定制
AlreadyKnown                    | [商业] 已学配方着色
AspectPosionBar                 | [多玩] 职业守护条
Atlas                           | [地图] Atlas [核心]
Atlas_Battlegrounds             | [地图] Atlas [战场地图]
Atlas_DungeonLocs               | [地图] Atlas [副本分布图]
Atlas_OutdoorRaids              | [地图] Atlas [户外首领分布图]
Atlas_Transportation            | [地图] Atlas [飞行路线图]
AtlasLoot                       | [地图] AtlasLoot [核心]
AtlasLoot_BurningCrusade        | [地图] AtlasLoot [燃烧的远征]
AtlasLoot_Crafting              | [地图] AtlasLoot [交易技能]
AtlasLoot_OriginalWoW           | [地图] AtlasLoot [经典旧世]
AtlasLoot_WorldEvents           | [地图] AtlasLoot [户外首领]
AtlasLoot_WrathoftheLichKing    | [地图] AtlasLoot [巫妖王之怒]
AtlasQuest                      | [地图] AtlasQuest [副本任务]
Auctionator                     | [商业] 拍卖行助手
BattleInfo                      | [多玩] 战场指示器
BlinkHealthText                 | [多玩] 简易信息
BuffTimer                       | [多玩] 魔法增益
BuyEmAll                        | [辅助] 材料购买助手
CastDelayBar                    | [多玩] 施法延迟条
ChannelClean                    | [多玩] 刷屏过滤
ChatPopo                        | [多玩] 聊天泡泡
ClearFont                       | [界面] 游戏字体
Clique                          | [多玩] 一键施法
CloseUp                         | [界面] 试衣间
DBM-Core                        | Deadly Boss Mods 核心
DBM-ChamberOfAspects            | <DBM> 龙眠神殿
DBM-Coliseum                    | <DBM> 十字军的试炼
DBM-EyeOfEternity               | <DBM> 永恒之眼
DBM-GUI                         | <DBM> 设置界面
DBM-Icecrown                    | <DBM> 冰冠堡垒
DBM-Naxx                        | <DBM> 纳克萨玛斯
DBM-Onyxia                      | <DBM> 奥妮克希亚
DBM-Party-WotLK                 | <DBM> 队伍-WotLK
DBM-PvP                         | <DBM> PvP
DBM-SpellTimers                 | <DBM> 冷却助手
DBM-Ulduar                      | <DBM> 奥杜尔
DBM-VoA                         | <DBM> 阿尔卡冯的宝库
DBM-WorldEvents                 | <DBM> 世界事件
dct                             | [多玩] 战斗指示器 DCT
dct_damage                      | [多玩] 战斗指示器-伤害模块
dct_options                     | [多玩] 战斗指示器 设置模块
dct_spellAlert                  | [多玩] 战斗指示器-法术预警
Decursive                       | [多玩] 一键驱散
Duowan                          | [多玩] 多玩基本库
DuowanBar                       | [多玩] 多玩动作条
DuowanChat                      | [多玩] 多玩聊天
DuowanMove                      | [多玩] 多玩界面
DuowanQuest                     | [多玩] 任务增强
DuowanRuneBar                   | [多玩] 死亡骑士符文条
DuowanTradeskill                | [多玩] 制作增强
DWDKP                           | [多玩] 多玩DKP
DWGKP                           | [多玩] 多玩金团助手
EN_UnitFrames                   | [多玩] 头像增强
eqgp                            | 贡献点统计插件 (EPGP)
EquipSwitcher                   | [多玩] 快速换装
EventAlert                      | [多玩] 技能特效报警
<del> FixItemIcon               | 问号物品修复 - 熊熊魔兽 </del> (已删除)
Fizzle                          | [多玩] 装备耐久
FriendsMenuXP                   | [多玩] 右键菜单增强
GatherHud                       | 采集助手 HUD
GatherMate                      | [多玩] 采集助手
GatherMate_Data                 | [多玩] 采集助手数据库
GearScoreLite                   | [多玩] 装备评分
GladiatorlosSA                  | [多玩] 竞技语音
Gladius                         | [多玩] 竞技场大师
GoGoMount                       | 随机坐骑
Grid                            | [多玩] 团队框架
GTFO                            | 危险站位提醒
HunterMod                       | [多玩] 猎人助手
InstanceMaps                    | [多玩] 副本地图
JPack                           | [多玩] 背包整理
LearningAid                     | [多玩] 法术助手 
Love                            | [多玩] 爱心工具
Ludwig                          | 物品数据库
MapNotes                        | [多玩] 地图标记
Mapster                         | [多玩] 地图增强
Mendeleev                       | [多玩] 物品制造助手
MerchantAssist                  | [多玩] 商业助手
NPCMark                         | NPC 标记
Omen                            | [多玩] 仇恨统计
OneBag3                         | 背包整合
OneBank3                        | 银行整合
OptiTaunt                       | [多玩] 嘲讽抵抗通报
oRA3                            | [多玩] 团队助手
Overachiever                    | 成就面板增强工具
Overachiever_Tabs               | 成就面板增强工具 - 标签
Overachiever_Trade              | 成就面板增强工具 - 商业
PallyPower                      | [多玩] 祝福管理
PhoenixStyle                    | [多玩] 团长助手 
PhoenixStyleMod_Coliseum        | [多玩] 团长助手 - 十字军模块
PhoenixStyleMod_Icecrown        | [多玩] 团长助手 - 巫妖王模块
PhoenixStyleMod_Ulduar          | [多玩] 团长助手 - 奥杜尔模块
Postal                          | [多玩] 邮件助手
PowerSource                     | [辅助] 法术来源
Quartz                          | 自定义施法条
QuestInfo                       | [多玩] 任务查询
RaidAchievement                 | 副本成就辅助 - 核心
RaidAchievement_Icecrown        | 副本成就辅助 - 冰冠冰川
RaidAchievement_Naxxramas       | 副本成就辅助 - 纳克萨玛斯
RaidAchievement_Ulduar          | 副本成就辅助 - 奥杜尔
RaidAchievement_WotlkHeroics    | 副本成就辅助 - 巫妖王地下城
RaidAlerter                     | [多玩] 团队报警系统
RangeDisplay                    | [多玩] 距离显示
RatingBuster                    | [多玩] 装备属性分析
RatingSummary                   | [多玩] 装备属性合计
Recount                         | [多玩] 伤害统计
RelationInfo                    | [多玩] YY好友助手
SharedMedia                     | [界面] 界面材质包
SharedMediaAdditionalFonts      | [界面] 界面材质包额外字体
Skillet                         | [商业] 商业技能助手
SpellID                         | 法术 ID 显示
SpellTimer                      | [多玩] 法术计时器
tdCooldown2                     | [多玩] 冷却计时
TexBrowser                      | [界面] 纹理浏览器
TheBurningTrade                 | [多玩] 交易增强
TidyPlates                      | [界面] 姓名版增强 (含连击点和减益显示)
TidyPlates_Neon                 | [界面] 姓名版增强 - Neon 风格
TipTac                          | [界面] 鼠标提示
TipTacItemRef                   | [界面] 鼠标提示 - 物品提示
TipTacOptions                   | [界面] 鼠标提示 - 选项
TipTacTalents                   | [界面] 鼠标提示 - 天赋
Titan                           | [多玩] 泰坦信息条
TotemTimers                     | [多玩] 图腾助手
TradeskillInfo                  | [多玩] 专业技能信息库
TradeskillInfoFu                | [多玩] 商业技能资料库 - 小地图按钮
TradeskillInfoUI                | [多玩] 商业技能信息库 - 设置
TrinketMenu                     | [多玩] 饰品管理
WhisperPop                      | [多玩] 密语聊天
XBar                            | XBar 显示额外的技能/法术按钮界面
XTrackBar                       | XBar 追踪模块
yClassColors                    | [界面] 姓名着色
YssBossLoot                     | [多玩] 首领掉落



CHANGELOG.md
# Change Log
基于多玩魔盒 3.3.5, 添加自用单体插件整合而成

## 2018-08-06
* 增加 ActionBar_Customer 动作条定制插件, 用来美化动作条快捷键文本
* 定制如下 <动作条快捷键默认字体为: Fonts\FRIZQT__.TTF 字体>: 
* Alt, Ctrl, Shift 快捷键转为大写(如A-1, C-E, 等)
* 鼠标左键,右键定制为: M-L(左), M-R(右)
* 鼠标中键及附加按键定制为: M-3 类似 M-数字 的显示方式
* 鼠标滚轮定制为: M-U(上滚), M-D(下滚)
* 注意: 因快捷键只能完整显示3个字符, 如果超过, 则显示2个字符+省略号, 因此, 不推荐鼠标+(Alt, Ctrl, Shift等)组合键

## 2018-07-30
* 修复中文客户端时, DBM 工作不正常的问题

## 2018-07-28
* 添加 贡献统计 GPGP 插件
* 尝试修复 DBM

## 2018-07-17
* 更新 GatherMate
* 添加 GatherHud 插件(屏幕中央显示追踪分布 HUD)
* 修复 BuyEmAll 某些情况下工作不正常的问题

## 2018-07-15
* 修复 TidyPlates 翻译警告
* 更换 DBM 为 Sunwell.pl 官方版本(与curseforge版本一致) - RAID 尚未进行测试

## 2018-07-12
* 添加姓名版增强插件 TidyPlates 并完成对其汉化
* 移除姓名版增益减益插件 PlateBuffs (功能与TidyPlates功能重叠)

## 2018-07-10
* 重新调整任务追踪显示, 文字超过长度将省略
* 加速 JPack 整理速度(未测试延迟高时的表现)
* 调整宠物动作条的位置 (仅测试 LR 和 DK)

## 2018-06-25
* 移除 "问号物品修复" 插件, 因其会导致战斗中拖动动作栏图标会导致图标不显示的问题
* 修正 EN_UnitFrames 错误提示
* 更新 Skillet 部分翻译, 及默认设置
* 调整 Skillet, 收支统计, 交易助手在山人页面的按钮布局

## 2018-06-24
* 添加 PlateBuffs 插件, 在姓名板显示 buff 或 debuff
* 设置 TexBrowser 默认情况下不进行加载 (LoadOnDemand)

## 2018-06-23
* 修复 ARL 鼠标提示BUG
* 调整 ARL 鼠标提示 (缩小提示宽度, 调整字体大小)
* 设置 GS 插件默认关闭对比
* 添加 XBar 插件及追踪模块, 用于方便快速手动切换追踪

## 2018-06-22
* 将 ARL 插件的鼠标提示默认设置为鼠标跟随

## 2018-06-20
* 降低 JPack 物品整理速度, 减少物品被锁定的几率
* 修正 GS插件部分 Bug 提示
* 修正任务追踪框架(WatchFrame)某些情况下过宽的问题

## 2018-06-13
* 修复GS插件部分时候(主要是鼠标指向单位框体时)不显示的问题
* 修改鼠标提示背景色为不透明

## 2018-06-12
* 重新修复任务追踪任务标题文字重叠bug

## 2018-06-11
* 法术计时器监视列表为死亡骑士添加"孤寂"BUFF
* 修复装备比较时, 当目标为传家宝时, 鼠标提示重叠问题 (未修复重复提示问题)

## 2018-06-10
* 更新背包整理插件 JPack
* 为邮箱插件增加快速删除按钮等模块
* 修复头像框不时上移的bug (主要出现在切换界面,进组,退组等情况)
* 修复任务追踪任务标题文字重叠bug
* 修复当打开世界地图时, 任务窗口详情被清空的bug
* 默认禁用采集助手
* 已知遗留问题:
  * 当开启采集助手后, 打开世界地图时, 严重影响 FPS, 请等待后续修复
  * 当登录游戏后, 首次打开任务面板后直接点击右上角地图按钮显示地图时, 任务详情面板内容会被清空, 此BUG属于暴雪原有问题, 暂不进行修复
  * 使用背包整理时, 仍会出现整理卡死状况, 大脚版本的背包整理暂时缓解, 此版本尚未进行此方面的优化
  * 重载界面后, 如果当前有宝宝, 重载后宝宝动作条中所有图标将会消失, 并无法使用, 此为暴雪原有BUG, 暂未修复, 临时请重新召唤或者上下坐骑即可恢复

## 2018-06-09
* 修复部分bug
* 整体调节各插件配合度
* 修复背包整理插件不能整理银行的问题

## 2018-06-03
* 调整一系列默认设置
* 跟新部分插件至最新版

## 2018-06-03
* 创建插件包


AddOns/Love/Love.lua
------------------------------------------------------------------------------------------
-- Love ver 1.0
-- 日期: 2010-3-29
-- 作者: dugu
-- 描述: 游戏中我们也有真心的朋友, 鱼别丢了, 我永远记得你; 现实中我们有自己心爱的人,
-- 愿与之生死相随; 愿大家珍惜这份友情和爱情, 愿世间充满爱.
-- 版权所有 (c) duowan.com
------------------------------------------------------------------------------------------
LoveDB = {};
local ident;
local palyerName;
local playerFaction;
local curRealm;
-----------------
-- 记录好友信息, 并可以导入小号
local db;
local PlayerList = {};
local IMPORT_FRIEND_TEXT = "";
if (GetLocale() == "zhCN") then
    IMPORT_FRIEND_LABEL = "确定将|cFF00FF00>>%s<<|R的好友导入当前角色";
    LOVE_SELECT_NAME = "选择名字";
    LOVE_SET_FOCUS = "◆Shift - 左键设置焦点";
    LOVE_CLEAR_FOCUS = "◆Shift - 右键取消焦点";
    LOVE_FRIEND_LOG_TEXT = "爱心工具";
    LOVE_FRIEND_LOG_TIP = "将所选角色的好友导入当前角色";
    LOVE_FRIEND_LOG_NONE = "没有 > %s < 的好友信息";
    LOVE_IMPORT_BUTTON_TEXT = "导";
    LOVE_FRIEND_NO_SELECT = "请选择一个角色";
elseif (GetLocale() == "zhTW") then
    IMPORT_FRIEND_LABEL = "確定將|cFF00FF00>>%s<<|R的好友導入當前角色";
    LOVE_SELECT_NAME = "選擇名字";
    LOVE_SET_FOCUS = "Shift - 左鍵設置焦點";
    LOVE_CLEAR_FOCUS = "Shift - 右鍵取消焦點";
    LOVE_FRIEND_LOG_TEXT = "愛心工具";
    LOVE_FRIEND_LOG_TIP = "將所選角色的好友導入當前角色";
    LOVE_FRIEND_LOG_NONE = "沒有 > %s < 的好友資訊";
    LOVE_IMPORT_BUTTON_TEXT = "導";
    LOVE_FRIEND_NO_SELECT = "請選擇一個角色";
else
    IMPORT_FRIEND_LABEL = "确定将|cFF00FF00>>%s<<|R的好友导入当前角色";
    LOVE_SELECT_NAME = "选择名字";
    LOVE_SET_FOCUS = "Shift - 左键设置焦点";
    LOVE_CLEAR_FOCUS = "Shift - 右键取消焦点";
    LOVE_FRIEND_LOG_TEXT = "love Tool";
    LOVE_FRIEND_LOG_TIP = "Click to inport frinds information";
    LOVE_FRIEND_LOG_NONE = "None infomation about this charactor > %s <";
    LOVE_IMPORT_BUTTON_TEXT = "IM";
    LOVE_FRIEND_NO_SELECT = "请选择一个角色";
end

local frame = CreateFrame("Frame");
frame:RegisterEvent("FRIENDLIST_UPDATE");
frame:RegisterEvent("ADDON_LOADED");

StaticPopupDialogs["IMPORT_FRIEND"] = {
    text = IMPORT_FRIEND_LABEL,
    button1 = OKAY,
    button2 = CANCEL,
    OnAccept = function(self)
        local name = UIDropDownMenu_GetText(ImPortFriendDropDown);
        ReAddFriends(name);
        self:Hide();
    end,
    showAlert = 1,
    timeout = 0,
    exclusive = 1,
    whileDead = 1,
    hideOnEscape = 1
};

local function canAdd(name)
    for i, n in ipairs(db) do
        if (n == name) then
            return false;
        end
    end
    return true;
end

local function SaveFriendsInfo()
    local name, level, class, area, connected, status, note;
    local maxNum = GetNumFriends();
    for index = 1, maxNum do
        name, level, class, area, connected, status, note = GetFriendInfo(index);
        if (canAdd(name)) then
            table.insert(db, name);
        end
    end
end

local function GetIdentInfo(id)
    local releam, faction, name = strsplit("_", id);
    return releam, faction, name;
end

local function UpdatePlayerNames()
    local releam, faction, name;
    for id, v in pairs(LoveDB) do
        releam, faction, name = GetIdentInfo(id);
        if (releam == curRealm and faction == playerFaction and name ~= palyerName) then
            tinsert(PlayerList, name);
        end
    end
end

function ReAddFriends(name)
    assert(name and type(name) == "string", "Argument #1 must be a string value.");

    local id = name and string.format("%s_%s_%s", curRealm, playerFaction, name) or ident;
    local db = LoveDB[id];
    if (not db or db == {}) then
        print(format(LOVE_FRIEND_LOG_NONE, tostring(id)));
        return;
    end

    for i, name in ipairs(db.FriendsLogDB) do
        AddFriend(name);
    end
end

local dropdown = CreateFrame("Frame", "ImPortFriendDropDown", FriendsListFrame, "UIDropDownMenuTemplate");
--dropdown:SetPoint("BOTTOMLEFT", IgnoreFrameToggleTab2, "BOTTOMRIGHT", -12, -6);
dropdown:SetPoint("BOTTOMLEFT", FriendsFrameStatusDropDown, "BOTTOMRIGHT", 120, -28);
UIDropDownMenu_SetWidth(dropdown, 26);
UIDropDownMenu_SetButtonWidth(dropdown, 24);
UIDropDownMenu_JustifyText(dropdown, "LEFT");
local name, fontHeight, flags = _G[dropdown:GetName() .. "Text"]:GetFont();
_G[dropdown:GetName() .. "Text"]:SetFont(name, 12, flags);
UIDropDownMenu_SetText(dropdown, LOVE_SELECT_NAME);

local dText = dropdown:CreateFontString(nil, "OVERLAY");
dText:SetFont(DUOWAN_CHANGELOG_FONT, 11, "OUTLINE");
dText:SetJustifyH("CENTER");
dText:SetTextColor(1, 1, 1);
dText:SetPoint("LEFT", dropdown, "LEFT", 22, 2);
dText:SetText("●");

local button = CreateFrame("Button", nil, FriendsListFrame, "UIPanelButtonTemplate");
button:RegisterForDrag("LeftButton");
button:SetWidth(25);
button:SetHeight(25);
button:SetPoint("LEFT", dropdown, "RIGHT", -15, 2);
button:EnableMouse(true);
button:SetMovable(true);
button:SetText(LOVE_IMPORT_BUTTON_TEXT);
button:SetScript("OnClick", function(self)
    local name = UIDropDownMenu_GetText(dropdown);
    if (LOVE_SELECT_NAME == name) then
        UIErrorsFrame:AddMessage(LOVE_FRIEND_NO_SELECT, 1, 0.2, 0);
    else
        IMPORT_FRIEND_TEXT = format(IMPORT_FRIEND_LABEL, name);
        StaticPopupDialogs["IMPORT_FRIEND"].text = IMPORT_FRIEND_TEXT;
        StaticPopup_Show("IMPORT_FRIEND");
    end
end);
button:SetScript("OnEnter", function(self)
    GameTooltip:SetOwner(self, "ANCHOR_RIGHT");
    GameTooltip:SetText(LOVE_FRIEND_LOG_TEXT);
    GameTooltip:AddLine(LOVE_FRIEND_LOG_TIP, 1, 1, 1);
    GameTooltip:Show();
end);
button:SetScript("OnLeave", function(self)
    GameTooltip:Hide();
end);
button:SetScript("OnDragStart", function(self)
    self:StartMoving();
end);
button:SetScript("OnDragStop", function(self)
    self:StopMovingOrSizing();
end);

frame:SetScript("OnEvent", function(self, event, name)
    if (event == "ADDON_LOADED" and name == "Love") then
        self:UnregisterEvent("ADDON_LOADED");
        palyerName = UnitName("player");
        curRealm = GetCVar("realmName");
        playerFaction = UnitFactionGroup("player");

        ident = string.format("%s_%s_%s", curRealm, playerFaction, palyerName);
        LoveDB = LoveDB or {};
        LoveDB[ident] = LoveDB[ident] or {};
        LoveDB[ident].FriendsLogDB = LoveDB[ident].FriendsLogDB or {};
        db = LoveDB[ident].FriendsLogDB;
        SaveFriendsInfo();
        UpdatePlayerNames();

        UIDropDownMenu_Initialize(dropdown, function()
            local info = UIDropDownMenu_CreateInfo();
            for i = 1, #(PlayerList) do
                info.text = PlayerList[i];
                info.func = function(self)
                    UIDropDownMenu_SetSelectedID(dropdown, self:GetID());
                end;
                info.checked = nil;
                UIDropDownMenu_AddButton(info);
            end
        end);

        if (#(PlayerList) == 0) then
            UIDropDownMenu_DisableDropDown(dropdown);
            button:Disable();
        end
    elseif (event == "FRIENDLIST_UPDATE") then
        SaveFriendsInfo();
    end
end)

-- 接口
function Love_ImportFriend_Toggle(switch)
    if (switch) then
        button:Show();
        dropdown:Show();
    else
        button:Hide();
        dropdown:Hide();
    end
end

-----------------------------
-- 便捷焦点功能

local focusenabled = false;
local focusshowtip = false;
function Love_UnitFrame_OnEnter(self)
    if (focusenabled and focusshowtip) then
        if (self:GetName() == "TargetFrame") then
            GameTooltip:AddLine(LOVE_SET_FOCUS, 0, 0.68, 0.85);
        elseif (self:GetName() == "TgFocusFrame" or self:GetName() == "FocusFrame") then
            GameTooltip:AddLine(LOVE_CLEAR_FOCUS, 0, 0.68, 0.85);
        end

        GameTooltip:Show();
    end
end

hooksecurefunc("UnitFrame_OnEnter", Love_UnitFrame_OnEnter);
TargetFrame:HookScript("OnEnter", Love_UnitFrame_OnEnter);
FocusFrame:HookScript("OnEnter", Love_UnitFrame_OnEnter);

FocusFrame:HookScript("OnDragStart", function(self)
    if (FOCUS_FRAME_LOCKED) then
        if (IsShiftKeyDown()) then
            self:StartMoving();
        end
    end
end);

FocusFrame:HookScript("OnDragStop", function(self)
    self:StopMovingOrSizing();
    ValidateFramePosition(self, 25);
end);

function Love_QuickFocus_Toggle(switch)
    if (switch) then
        TargetFrame:SetAttribute("shift-type1", "focus");
        --TgFocusFrame:SetAttribute("shift-type2", "macro");
        --TgFocusFrame:SetAttribute("macrotext", "/clearfocus");
        FocusFrame:SetAttribute("shift-type2", "macro");
        FocusFrame:SetAttribute("macrotext", "/clearfocus");
        focusenabled = true;
    else
        TargetFrame:SetAttribute("shift-type1", "");
        --TgFocusFrame:SetAttribute("shift-type2", "");
        --TgFocusFrame:SetAttribute("macrotext", "");
        FocusFrame:SetAttribute("shift-type2", "");
        FocusFrame:SetAttribute("macrotext", "");
        focusenabled = false;
    end
end

function Love_QuickFocus_ShowTip(switch)
    if (switch) then
        focusshowtip = true;
    else
        focusshowtip = false;
    end
end

------------------------
-- 双采功能
local enableshuangcai = false;
local tracking_time = 0;
local tracking_count = 0;
local DW_FINDING_HERB, DW_FINDING_MINING;
local tracking_herb, tracking_mining;
local curtacking = "mining";
if (GetLocale() == "zhCN") then
    DW_FINDING_HERB = "寻找草药";
    DW_FINDING_MINING = "寻找矿物";
elseif (GetLocale() == "zhTW") then
    DW_FINDING_HERB = "尋找草藥";
    DW_FINDING_MINING = "尋找礦物";
else
    DW_FINDING_HERB = "Find Herbs";
    DW_FINDING_MINING = "Find Minerals";
end

local frame = CreateFrame("Frame", nil, UIParent);
local function Love_ShuangCai_OnUpdate(self, elapsed)
    if (enableshuangcai) then
        tracking_time = tracking_time + elapsed;
        if (tracking_time > 3) then
            tracking_time = 0;

            local count = GetNumTrackingTypes();
            if (tracking_count ~= count) then
                local name;
                tracking_herb, tracking_mining = false, false;
                for i = 1, count do
                    name = GetTrackingInfo(i);
                    if (name == DW_FINDING_HERB) then
                        tracking_herb = i;
                    elseif (name == DW_FINDING_MINING) then
                        tracking_mining = i;
                    end
                end
                tracking_count = count;
            end

            if (tracking_herb and tracking_mining) then
                if (curtacking == "mining") then
                    local start, duration, enabled = GetSpellCooldown(DW_FINDING_MINING);
                    local spell = UnitCastingInfo("player");
                    local combat = UnitAffectingCombat("player");
                    if (not (combat or spell or (start and duration and start > 0 and duration > 0))) then
                        SetTracking(tracking_mining);
                        curtacking = "herb";
                    end
                else
                    local start, duration, enabled = GetSpellCooldown(DW_FINDING_HERB);
                    local spell = UnitCastingInfo("player");
                    local combat = UnitAffectingCombat("player");
                    if (not (combat or spell or (start and duration and start > 0 and duration > 0))) then
                        SetTracking(tracking_herb);
                        curtacking = "mining";
                    end
                end
            end
        end
    end
end

frame:SetScript("OnUpdate", Love_ShuangCai_OnUpdate);

function Love_ShuangCai_Toggle(switch)
    enableshuangcai = switch;
end
